lib/handlers/authenticate-handler.js

```diff
@@ -0,0 +1,255 @@
+
+  return Promise.bind(this)
+    .then(function() {
+      return this.getToken(request);
+    })
+    .then(function(token) {
+      return this.getAccessToken(token);
+    })
+    .tap(function(token) {
+      return this.validateAccessToken(token);
+    })
+    .tap(function(token) {
+      if (!this.scope) {
+        return;
+      }
+
+      return this.validateScope(token);
+    })
+    .tap(function(token) {
+      return this.updateResponse(response, token);
+    })
+    .catch(function(e) {
+      // Include the "WWW-Authenticate" response header field if the client
+      // lacks any authentication information.
+      //
+      // @see https://tools.ietf.org/html/rfc6750#section-3.1
+      if (e instanceof UnauthorizedRequestError) {
+        response.set('WWW-Authenticate', 'Bearer realm="Service"');
+      }
+
+      if (!(e instanceof OAuthError)) {
+        throw new ServerError(e);
+      }
+
+      throw e;
+    });
+};
+
+/**
+ * Get the token from the header or body, depending on the request.
+ *
+ * "Clients MUST NOT use more than one method to transmit the token in each request."
+ *
+ * @see https://tools.ietf.org/html/rfc6750#section-2
+ */
+
+AuthenticateHandler.prototype.getToken = function(request) {
+  var headerToken = request.get('Authorization');
+  var queryToken = request.query.access_token;
+  var bodyToken = request.body.access_token;
+
+  if (!!headerToken + !!queryToken + !!bodyToken > 1) {
+    throw new InvalidRequestError('Invalid request: only one authentication method is allowed');
+  }
+
+  if (headerToken) {
+    return this.getTokenFromRequestHeader(request);
+  }
+
+  if (queryToken) {
+    return this.getTokenFromRequestQuery(request);
+  }
+
+  if (bodyToken) {
+    return this.getTokenFromRequestBody(request);
+  }
+
+  throw new UnauthorizedRequestError('Unauthorized request: no authentication given');
+};
+
+/**
+ * Get the token from the request header.
+ *
+ * @see http://tools.ietf.org/html/rfc6750#section-2.1
+ */
+
+AuthenticateHandler.prototype.getTokenFromRequestHeader = function(request) {
+  var token = request.get('Authorization');
+  var matches = token.match(/Bearer\s(\S+)/);
+
+  if (!matches) {
+    throw new InvalidRequestError('Invalid request: malformed authorization header');
+  }
+
+  return matches[1];
+};
+
+/**
+ * Get the token from the request query.
+ *
+ * "Don't pass bearer tokens in page URLs:  Bearer tokens SHOULD NOT be passed in page
+ * URLs (for example, as query string parameters). Instead, bearer tokens SHOULD be
+ * passed in HTTP message headers or message bodies for which confidentiality measures
+ * are taken. Browsers, web servers, and other software may not adequately secure URLs
+ * in the browser history, web server logs, and other data structures. If bearer tokens
+ * are passed in page URLs, attackers might be able to steal them from the history data,
+ * logs, or other unsecured locations."
+ *
+ * @see http://tools.ietf.org/html/rfc6750#section-2.3
```

Comment:
I absolutely understand your point and I support adding this behaviour but I suggest adding it as an optional flag, defaulting to disabled.

The reason is that I consider this to be extremely insecure. By passing a token as a query string you risk leaking it (through a referrer header for example) to dependencies of your application (google analytics, an embedded iframe, any sort of script really).

Any thoughts?